File: easy/3903-smallest-stable-index-i/solution.py
from typing import List


class Solution:
    def smallestStableIndex(self, nums: List[int], k: int) -> int:
        """
        3903. Smallest Stable Index I
        Time: O(n)
        Space: O(n)
        """
        n = len(nums)

        # suf_min[i] = min(nums[i..n-1])
        suf_min = [0] * n
        suf_min[n - 1] = nums[n - 1]
        for i in range(n - 2, -1, -1):
            suf_min[i] = min(nums[i], suf_min[i + 1])

        # Running prefix max; scanning left to right finds the smallest index.
        pref_max = float("-inf")
        for i in range(n):
            pref_max = max(pref_max, nums[i])
            if pref_max - suf_min[i] <= k:
                return i
        return -1

File: easy/3903-smallest-stable-index-i/solution.java
/**
 * 3903. Smallest Stable Index I
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int smallestStableIndex(int[] nums, int k) {
        int n = nums.length;

        // sufMin[i] = min(nums[i..n-1])
        int[] sufMin = new int[n];
        sufMin[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            sufMin[i] = Math.min(nums[i], sufMin[i + 1]);
        }

        // Running prefix max; scanning left to right finds the smallest index.
        int prefMax = Integer.MIN_VALUE;
        for (int i = 0; i < n; i++) {
            prefMax = Math.max(prefMax, nums[i]);
            if (prefMax - sufMin[i] <= k) {
                return i;
            }
        }
        return -1;
    }
}
